package.json

```diff
@@ -64,6 +64,7 @@
   ],
   "sideEffects": false,
   "dependencies": {
+    "@graphql-typed-document-node/core": "^3.1.0",
```

Comment:
@longility, do you think there's a way to support typed document nodes without introducing this as a dependency?

Not all the users will use GraphQL, and even smaller subset of those who do will use `TypedDocumentNode`. Adding this as a fixed dependency is quite a price _everybody_ will pay.

From the type definitions, `TypedDocumentNode` and `DocumentNode` seem compatible. The only issue I see is that one cannot extract the query fields from `DocumentNode` straight away, while `TypedDocumentNode` has generics to infer them in a handler. 

Do you think we could swap `TypedDocumentNode` with the plain `DocumentNode` (users already install `graphql`)? We can write a custom extract type to get the query fields from the plain document node to achieve type safety, I think.

Reply:
We can still use this package as a dev dependency, so we could generate some typed document nodes and use them in type tests, for example. But I don't think this is something our users should install.

Reply:
Yes, I had concerns with this but I did not know how to approach. One thought I had was that we can duplicate the `TypedDocumentNode` contract into the main package code. 

```js
export interface TypedDocumentNode<Result = {
    [key: string]: any;
}, Variables = {
    [key: string]: any;
}> extends DocumentNode {
    __resultType?: Result;
    __variablesType?: Variables;
}
```

Maybe add `@graphql-typed-document-node/core` as dev dependency for testing. What do you think of that approach? I assume `TypedDocumentNode` contract shouldn't change too often and an acceptable risk, but I'm just guessing.

Reply:
@kettanaito I updated the code to where there isn't a dependency on `@graphql-typed-document-node/core`, and duplicated the contract. I figure easier to see the code. I'm not sure how to test. You would have to guide me as I'm weak in that area. I decided not to add `@graphql-typed-document-node/core` as a dev dependency unless we are going to test.

Reply:
@kettanaito I was able to figure out and test following the pattern in place with `graphql.test-d.ts` and having `graphql-typed-document-node/core` as a dev dependency. Let me know what you think.

Reply:
Hey, @longility. Those changes look fantastic. 
Could you please tell me how do you generate the `graphql.test-data.ts`?

Reply:
Of course.

Here is an example. I actually used this to generate the `graphql.test-data.ts`. https://github.com/longility/msw/commit/7dc85f9a0d7f0669b469884f97b96b231f997b68

I used this for reference: https://github.com/dotansimha/graphql-typed-document-node#how-to-use

Reply:
Of course.

Here is an example. I actually used this to generate the `graphql.test-data.ts`. https://github.com/longility/msw/commit/b96da417cbab88d9c6724dae2abb25cede72b36b

I used this for reference: https://github.com/dotansimha/graphql-typed-document-node#how-to-use

Know that you can run against like `gql` tag function. In example, I'm using `*.graphql` file instead.

I have not done enough to know how I feel about all this yet as mentioned before. I'm experimenting. :)

Reply:
The more I think about this feature, the more it sounds like a standalone usage example. I'm thinking maybe we can move this into a `with-graphql-code-generator` example in our examples repo, while leaving the support implementation in MSW. What are your thoughts on this? Let's discuss this before proceeding if you don't mind.

Reply:
I don't mind @kettanaito . I want to consider your input more heavily as you are more familiar with the repo and ecosystem of msw than I am.

If you are talking about the codegen example commit I linked (which I used to generate `graphql.test-data.ts`), then yes, I think it makes sense for `with-graphql-code-generator` in https://github.com/mswjs/examples/tree/master/examples is what I believe you are talking about. Also, maybe an addition to the docs like somewhere here: https://mswjs.io/docs/api/graphql/query#usage-with-typescript.

For transparency, this PR does not have that codegen example commit I shared, that was another branch to help you get an idea and I didn't want to add that extra maintenance on this PR.